File: LeetCode 78 90-subset/unique_nums/subset_iteration.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        subsets = [[]]
        
        for num in nums:
            length = len(subsets)
            for i in range(length):
                subsets.append([num] + subsets[i])
        
        return subsets

if __name__ == "__main__":
    solution = Solution()
    nums = [1, 2, 3]
    print("subsets", solution.subsets(nums))

# output: ('subsets', [[], [1], [2], [2, 1], [3], [3, 1], [3, 2], [3, 2, 1]])

File: LeetCode 78 90-subset/unique_nums/subset_iteration.go
package main

import "fmt"

func subsets(nums []int) [][]int {
	subsets := [][]int{{}}

	for _, num := range nums {
		length := len(subsets)
		for i := 0; i < length; i++ {
			newSubset := []int{}
			newSubset = append([]int{num}, subsets[i]...)
			subsets = append(subsets, newSubset)
		}
	}

	return subsets
}

func main() {
	nums := []int{1, 2, 3}
	fmt.Println("subsets", subsets((nums)))
}

// output: subsets [[] [1] [2] [2 1] [3] [3 1] [3 2] [3 2 1]]
